File: m2cgen/interpreters/c_sharp/softmax.cs
private static double[] Softmax(double[] x) {
    int size = x.Length;
    double[] result = new double[size];
    double max = x[0];
    for (int i = 1; i < size; ++i) {
        if (x[i] > max)
            max = x[i];
    }
    double sum = 0.0;
    for (int i = 0; i < size; ++i) {
        result[i] = Exp(x[i] - max);
        sum += result[i];
    }
    for (int i = 0; i < size; ++i)
        result[i] /= sum;
    return result;
}

File: m2cgen/interpreters/fortran/softmax.f90
function softmax(x) result(res)
    implicit none
    double precision, dimension(:), intent(in) :: x
    double precision, dimension(size(x)) :: res
    double precision :: max_val, sum_val
    integer :: i

    ! Find maximum value in x
    max_val = x(1)
    do i = 2, size(x)
        if (x(i) > max_val) then
            max_val = x(i)
        end if
    end do

    ! Compute softmax values
    sum_val = 0.0d0
    do i = 1, size(x)
        res(i) = exp(x(i) - max_val)
        sum_val = sum_val + res(i)
    end do
    res = res / sum_val

end function softmax
